gnpy/core/science_utils.py

```diff
@@ -0,0 +1,842 @@
+        matrix_size = max(carriers, key=lambda x: getattr(x, 'channel_number')).channel_number
+        eta_matrix = np.zeros(shape=(matrix_size, matrix_size))
+
+        # SPM
+        if self.nli_params.verbose:
+            print(f'Start computing SPM on channel #{carrier_cut.channel_number}')
+        # SPM GGN
+        if 'ggn' in self.nli_params.nli_method_name.lower():
+            partial_nli = self._generalized_spectrally_separated_spm(carrier_cut)
+        # SPM GN
+        elif 'gn' in self.nli_params.nli_method_name.lower():
+            partial_nli = self._gn_analytic(carrier_cut, *[carrier_cut])
+        eta_matrix[cut_index, cut_index] = partial_nli / (carrier_cut.power.signal**3)
+
+        # XPM
+        for pump_carrier in carriers:
+            pump_index = pump_carrier.channel_number - 1
+            if not (cut_index == pump_index):
+                if self.nli_params.verbose:
+                    print(f'Start computing XPM on channel #{carrier_cut.channel_number} '
+                          f'from channel #{pump_carrier.channel_number}')
+                # XPM GGN
+                if 'ggn' in self.nli_params.nli_method_name.lower():
+                    partial_nli = self._generalized_spectrally_separated_xpm(carrier_cut, pump_carrier)
+                # XPM GGN
+                elif 'gn' in self.nli_params.nli_method_name.lower():
+                    partial_nli = self._gn_analytic(carrier_cut, *[pump_carrier])
+                eta_matrix[pump_index, pump_index] = partial_nli /\
+                                                     (carrier_cut.power.signal * pump_carrier.power.signal**2)
+        return eta_matrix
+
+    # Methods for computing GN-model
+    def _gn_analytic(self, carrier, *carriers):
+        """ Computes the nonlinear interference power on a single carrier.
+        The method uses eq. 120 from arXiv:1209.0394.
+        :param carrier: the signal under analysis
+        :param carriers: the full WDM comb
+        :return: carrier_nli: the amount of nonlinear interference in W on the carrier under analysis
+        """
+        alpha = self.fiber_params.alpha0() / 2
+        beta2 = self.fiber_params.beta2
+        gamma = self.fiber_params.gamma
+        length = self.fiber_params.length
+        effective_length = (1 - np.exp(-2 * alpha * length)) / (2 * alpha)
+        asymptotic_length = 1 / (2 * alpha)
+
+        g_nli = 0
+        for interfering_carrier in carriers:
+            g_interfearing = interfering_carrier.power.signal / interfering_carrier.baud_rate
+            g_signal = carrier.power.signal / carrier.baud_rate
+            g_nli += g_interfearing**2 * g_signal * self._psi(carrier, interfering_carrier)
+        g_nli *= (16.0 / 27.0) * (gamma * effective_length)**2 /\
+                 (2 * np.pi * abs(beta2) * asymptotic_length)
+        carrier_nli = carrier.baud_rate * g_nli
+        return carrier_nli
+
+    def _psi(self, carrier, interfering_carrier):
```

Comment:
This is already in `gnpy/core/elements.py`. I'm all for moving all "real computation" to a dedicated submodule, but I would prefer to avoid a state where stuff got duplicated.

My preference would be to move these methods one-by-one (or at least separately from the NLI solver as done in commit f9bd6310f182b62f9d373f4cbbba11b0af312b9d).